src/Gaffer/ScriptNode.cpp

```diff
@@ -810,13 +811,30 @@ void ScriptNode::plugSet( Plug *plug )
 	{
 		context()->setFramesPerSecond( framesPerSecondPlug()->getValue() );
 	}
+	else if( auto nameValuePlug = plug->parent<NameValuePlug>() )
+	{
+		if(
+			nameValuePlug->parent() == variablesPlug() &&
+			plug == nameValuePlug->enabledPlug() &&
+			!nameValuePlug->enabledPlug()->getValue()
+		)
+		{
+			// NameValuePlug has been disabled, so remove variable from context. There
+			// may be another NameValuePlug specifying a variable of the same name,
+			// which should take effect now this one is disabled. But we remove
+			// unconditionally and then leave the `plug == variablesPlug()` case below
+			// to add the variable back if necessary (`plugSet()` is emitted on the
+			// parent after being emitted for the child).
+			context()->remove( nameValuePlug->namePlug()->getValue() );
+		}
+	}
 	else if( plug == variablesPlug() )
 	{
 		IECore::CompoundDataMap values;
 		variablesPlug()->fillCompoundData( values );
-		for( IECore::CompoundDataMap::const_iterator it = values.begin(), eIt = values.end(); it != eIt; ++it )
+		for( const auto &variable : values )
```

Comment:
> If you delete the last character of the name from a NameValuePlug, that should remove the variable from the context. But we don't know what that last character was - in `plugSet()` that information has been lost already.

Could one way around this be to keep a runtime cache here of the list of variables names set here? Then the next time around, whenever there is a variable in that list that isn't in `values`, you know you can remove it? I think that would cover all the various enabled/removed cases too.

Reply:
Yep, I'd just been avoiding that as it meant new member data, and therefore `master` rather than `0.58_maintenance`. But this doesn't seem to be the most pressing fix, and it's certainly a lot simpler with your suggestion. I've updated the code, added tests for the renaming of variables (including renaming to "") and update the PR base to master.